File: docs/scripts_js/ARC.js
import {ARC} from "@nfdi4plants/arctrl";
import {fulfillWriteContract, fulfillReadContract} from "./Contracts.js";

// # Create

let arc = new ARC()

// # Write
const arcRootPath = "C:/Users/Kevin/Desktop/NewTestARCJS"

async function write(arcPath, arc)  {
    let contracts = arc.GetWriteContracts()
    for (const contract of contracts) {
        // from Contracts.js docs
        await fulfillWriteContract(arcPath,contract)
    };
}

await write(arcRootPath, arc)

// # Read

// Setup
function normalizePathSeparators (str) {
  const normalizedPath = path.normalize(str)
  return normalizedPath.replace(/\\/g, '/');
}

export function getAllFilePaths(basePath) {
  const filesList = []
  function loop (dir) {
      const files = fs.readdirSync(dir);
      for (const file of files) {
          const filePath = path.join(dir, file);
  
          if (fs.statSync(filePath).isDirectory()) {
              // If it's a directory, recursively call the function on that directory
              loop(filePath);
          } else {
              // If it's a file, calculate the relative path and add it to the list
              const relativePath = path.relative(basePath, filePath);
              const normalizePath = normalizePathSeparators(relativePath)
              filesList.push(normalizePath);
          }
      }
  }
  loop(basePath)
  return filesList;
}

// put it all together
async function read(basePath) {
    let allFilePaths = getAllFilePaths(basePath)
    // Initiates an ARC from FileSystem but no ISA info.
    let arc = ARC.fromFilePaths(allFilePaths)
    // Read contracts will tell us what we need to read from disc.
    let readContracts = arc.GetReadContracts()
    console.log(readContracts)
    let fcontracts = await Promise.all(
        readContracts.map(async (contract) => {
            let content = await fulfillReadContract(basePath, contract)
            contract.DTO = content
            return (contract) 
        })
    )
    arc.SetISAFromContracts(fcontracts)
    return arc
}

// Execution

await read(arcRootPath).then(arc => console.log(arc))

File: docs/scripts_python/ARC.py
import os
from arctrl.arc import ARC
from Contract import fulfill_write_contract, fulfill_read_contract
# Create

arc = ARC()

# Write
arc_root_path = "C:/Users/Kevin/Desktop/NewTestARCJS"

async def write(arc_path, arc: ARC):
    contracts = arc.GetWriteContracts()
    for contract in contracts:
        # from Contracts.js docs
        await fulfill_write_contract(arc_path, contract)

# Read

# Setup
def normalize_path_separators(path_str: str):
    normalized_path = os.path.normpath(path_str)
    return normalized_path.replace('\\', '/')

def get_all_file_paths(base_path):
    files_list = []
    def loop(dir_path):
        files = os.listdir(dir_path)
        for file_name in files:
            file_path = os.path.join(dir_path, file_name)
            if os.path.isdir(file_path):
                loop(file_path)
            else:
                relative_path = os.path.relpath(file_path, base_path)
                normalize_path = normalize_path_separators(relative_path)
                files_list.append(normalize_path)
    loop(base_path)
    return files_list

# put it all together
def read(base_path):
    all_file_paths = get_all_file_paths(base_path)
    arc = ARC.from_file_paths(all_file_paths)
    read_contracts = arc.GetReadContracts()
    print(read_contracts)
    fcontracts = (fulfill_read_contract(base_path, contract) for contract in read_contracts)
    for contract, content in zip(read_contracts, fcontracts):
        contract.DTO = content
    arc.SetISAFromContracts(fcontracts)
    return arc
